src/hotspot/share/opto/loopnode.cpp

```diff
@@ -3675,9 +3679,15 @@ void PhaseIdealLoop::replace_parallel_iv(IdealLoopTree *loop) {
         incr2->in(1)->uncast() != phi2 ||
         incr2 == incr ||
         incr2->Opcode() != Op_AddI ||
-        !incr2->in(2)->is_Con())
+        !incr2->in(2)->is_Con()) {
       continue;
+    }
 
+    if (incr2->in(1)->is_ConstraintCast() && !incr2->in(1)->in(0)->is_RangeCheck()) {
+      // Skip AddI->CastII->Phi case if CastII is not controlled by local RangeCheck
+      // to reflect changes in LibraryCallKit::inline_preconditions_checkIndex
```

Comment:
In the valid case, isn't the ConstraintCast control input `incr2->in(1)->in(0)` the IfTrue projection of the RangeCheck?

I would remove the second line because it's not clear which "changes" in `LibraryCallKit::inline_preconditions_checkIndex` it is referring to.

```suggestion
```

Reply:
It's still not clear to me how the `incr2->in(1)->in(0)->is_RangeCheck()` condition can ever be true. How can the control input of a ConstraintCast be a RangeCheck? Shouldn't there be a projection node in-between?

Reply:
I missed this comment before... I need incr2->in(1)->in(0)->in(0)->is_RangeCheck()